index.js

```diff
@@ -1,174 +1,162 @@
-      if (!this._encoded) this._encode()
-      return this._encoded
+  decode () {
+    const decoded = this.decodeUnsafe()
+    if (decoded === null) return null
+    if (isBinary(decoded)) return copyBinary(decoded)
+    if (immutableTypes.has(typeof decoded) || decoded === null) {
+      return decoded
     }
+    return clone(decoded)
+  }
 
-    _decode () {
-      if (typeof this.opts.source !== 'undefined') this._decoded = this.opts.source
-      else {
-        const { decode } = multicodec.get(this.code)
-        this._decoded = decode(this._encoded || this.opts.data)
-      }
-      return this._decoded
-    }
+  reader () {
+    return reader(this.decodeUnsafe())
+  }
 
-    decode () {
-      // TODO: once we upgrade to the latest data model version of
-      // dag-pb that @gozala wrote we should be able to remove this
-      // and treat it like every other codec.
-      /* c8 ignore next */
-      if (this.codec === 'dag-pb') return this._decode()
-      if (!this._decoded) this._decode()
-      const tt = typeof this._decoded
-      if (tt === 'number' || tt === 'boolean') {
-        // return any immutable types
-        return this._decoded
-      }
-      if (isBinary(this._decoded)) return copyBinary(this._decoded)
-      if (immutableTypes.has(typeof this._decoded) || this._decoded === null) {
-        return this._decoded
-      }
-      return clone(this._decoded)
-    }
+  async equals (block) {
+    if (block === this) return true
+    if (block.asBlock !== block) throw new Error('Not a block instance')
+    const [a, b] = await Promise.all([this.cid(), block.cid()])
+    return a.equals(b)
+  }
+}
 
-    decodeUnsafe () {
-      if (!this._decoded) this._decode()
-      return this._decoded
-    }
+Block.codecs = new Map()
 
-    reader () {
-      return reader(this.decodeUnsafe())
-    }
+Block.add = codec => {
+  if (codec.name) Block.codecs.set(codec.name, codec)
+  if (codec.code) Block.codecs.set(codec.code, codec)
+}
+Block.add(json)
+Block.add(raw)
 
-    async equals (block) {
-      if (block === this) return true
-      const cid = await this.cid()
-      if (CID.asCID(block)) return cid.equals(CID.asCID(block))
-      // https://github.com/bcoe/c8/issues/135
-      /* c8 ignore next */
-      return cid.equals(await block.cid())
-    }
+Block.encoder = (source, codec, hasher = sha256) => {
+  if (typeof codec === 'string') codec = Block.codecs.get(codec)
+  if (!codec) throw new Error('Missing codec')
+  return new Block({ source, codec, hasher })
+}
+Block.decoder = (data, codec, hasher = sha256) => {
+  if (typeof codec === 'string') codec = Block.codecs.get(codec)
+  if (!codec) throw new Error('Missing codec')
+  return new Block({ data, codec, hasher })
+}
+Block.createUnsafe = (data, cid, { hasher, codec } = {}) => {
+  codec = codec || Block.codecs.get(cid.code)
+  if (!codec) throw new Error(`Missing codec ${cid.code}`)
+  return new Block({ data, cid, codec, hasher: hasher || null })
```

Comment:
Requires change above.

```suggestion
  return new Block({ data, cid, codec })
```